lib/executor/ramping_vus.go

```diff
@@ -493,154 +493,194 @@ var _ lib.Executor = &RampingVUs{}
+	// the end of gracefulStop (= maxDur-regDur) timeouts
+	var (
+		handleNewMaxAllowedVUs = rs.maxAllowedVUsHandlerStrategy()
+		handleNewScheduledVUs  = rs.scheduledVUsHandlerStrategy()
+		wait                   = waiter(ctx, rs.started)
+	)
+	for len(rs.stepsRaw) > 0 {
+		var (
+			r, g   = rs.stepsRaw[0], rs.stepsGraceful[0]
+			gfirst = r.TimeOffset > g.TimeOffset
+			offset = g.TimeOffset
+		)
+		if !gfirst {
+			offset = r.TimeOffset
+		}
+		if wait(offset) {
+			return
+		}
+		if gfirst {
+			handleNewMaxAllowedVUs(g)
+			rs.stepsGraceful = rs.stepsGraceful[1:]
 		} else {
-			for vuNum := newScheduledVUs; vuNum < currentScheduledVUs; vuNum++ {
-				vuHandles[vuNum].gracefulStop()
-			}
+			handleNewScheduledVUs(r)
+			rs.stepsRaw = rs.stepsRaw[1:]
+		}
+	}
+}
+
+// TODO: removing this has no effect on tests?
+func (rs rampingVUsRunState) handleRemainingVUs(ctx context.Context) {
+	var (
+		handleNewMaxAllowedVUs = rs.maxAllowedVUsHandlerStrategy()
+		wait                   = waiter(ctx, rs.started)
+	)
+	for _, s := range rs.stepsGraceful {
+		if wait(s.TimeOffset) {
+			return
 		}
-		currentScheduledVUs = newScheduledVUs
+		handleNewMaxAllowedVUs(s)
 	}
+}
 
-	handleNewMaxAllowedVUs := func(newMaxAllowedVUs uint64) {
-		if newMaxAllowedVUs < currentMaxAllowedVUs {
-			for vuNum := newMaxAllowedVUs; vuNum < currentMaxAllowedVUs; vuNum++ {
-				vuHandles[vuNum].hardStop()
+func (rs rampingVUsRunState) maxAllowedVUsHandlerStrategy() func(lib.ExecutionStep) {
+	var max uint64
+	return func(graceful lib.ExecutionStep) {
+		cur := graceful.PlannedVUs
+		if cur < max {
+			for n := cur; n < max; n++ {
+				rs.vuHandles[n].hardStop()
 			}
 		}
-		currentMaxAllowedVUs = newMaxAllowedVUs
+		max = cur
 	}
+}
 
-	wait := waiter(parentCtx, startTime)
-	// iterate over rawExecutionSteps and gracefulExecutionSteps in order by TimeOffset
-	// giving rawExecutionSteps precedence.
-	// we stop iterating once rawExecutionSteps are over as we need to run the remaining
-	// gracefulExecutionSteps concurrently while waiting for VUs to stop in order to not wait until
-	// the end of gracefulStop timeouts
-	i, j := 0, 0
-	for i != len(rawExecutionSteps) {
-		if rawExecutionSteps[i].TimeOffset > gracefulExecutionSteps[j].TimeOffset {
-			if wait(gracefulExecutionSteps[j].TimeOffset) {
-				return
+func (rs rampingVUsRunState) scheduledVUsHandlerStrategy() func(lib.ExecutionStep) {
+	var sched uint64
+	return func(raw lib.ExecutionStep) {
+		cur := raw.PlannedVUs
+		if cur > sched {
+			for n := sched; n < cur; n++ {
+				_ = rs.vuHandles[n].start() // TODO handle error
 			}
-			handleNewMaxAllowedVUs(gracefulExecutionSteps[j].PlannedVUs)
-			j++
 		} else {
-			if wait(rawExecutionSteps[i].TimeOffset) {
-				return
+			for n := cur; n < sched; n++ {
+				rs.vuHandles[n].gracefulStop()
 			}
-			handleNewScheduledVUs(rawExecutionSteps[i].PlannedVUs)
-			i++
 		}
+		sched = cur
```

Comment:
I would:
1. make the `if` statement be the same as the for one .. .so reverse it as in `sched < cur`.
2. Given that we will set `sched` to `cur` there is no point in not just use the value of `sched` and increment/decrement it in the for loop instead of adding `n` as well.